src/porepy/models/run_convergence_study.py

```diff
@@ -0,0 +1,217 @@
+        params: dict,
+        n_refinements: int = 1,
+        newton_params: dict = None,
+        variable: List[str] = None,  # This is really required for the moment
+        variable_dof: List[int] = None,
+) -> Tuple[List[pp.GridBucket], List[dict]]:
+    """ Run a model on a grid, refined n times.
+
+    For a given model and method to run the model,
+    and a set of parameters, n refinements of the
+    initially generated grid will be performed.
+
+
+    Parameters
+    ----------
+    model : Type[ContactMechanics]
+        Which model to run
+    run_model_method : Callable
+        Which method to run model with
+        Typically pp.run_stationary_model or pp.run_time_dependent_model
+    network : Union[pp.FractureNetwork3d, pp.FractureNetwork2d]
+        Fracture network
+    params : dict (Default: None)
+        Custom parameters to pass to model
+    n_refinements : int (Default: 1)
+        Number of grid refinements
+    newton_params : dict (Default: None)
+        Any non-default newton solver parameters to use
+    variable : List[str]
+        List of variables to consider for convergence study.
+        If 'None', available variables on each sub-domain will be used
+    variable_dof : List[str]
+        Degrees of freedom for each variable
+
+    Returns
+    -------
+    gb_list : List[pp.GridBucket]
+        list of (n+1) grid buckets in increasing order
+        of refinement (reference grid last)
+    errors : List[dict]
+        List of (n) dictionaries, each containing
+        the error for each variable on each grid.
+        Each list entry corresponds index-wise to an
+        entry in gb_list.
+    """
+
+    # 1. Step: Create n grids by uniform refinement.
+    # 2. Step: for grid i in list of n grids:
+    # 2. a. Step: Set up the mechanics model.
+    # 2. b. Step: Solve the mechanics problem.
+    # 2. c. Step: Keep the grid (with solution data)
+    # 3. Step: Let the finest grid be the reference solution.
+    # 4. Step: For every other grid:
+    # 4. a. Step: Map the solution to the fine grid, and compute error.
+    # 5. Step: Compute order of convergence, etc.
+
+    logger.info(f"Preparing setup for convergence study on {pendulum.now().to_atom_string()}")
+
+    # 1. Step: Create n grids by uniform refinement.
+    gb_list = gb_refinements(
+        network=network,
+        gmsh_folder_path=params['folder_name'],
+        mesh_args=params['mesh_args'],
+        n_refinements=n_refinements,
+    )
+
+    # -----------------------
+    # --- SETUP AND SOLVE ---
+    # -----------------------
+    newton_params = newton_params if newton_params else {}
+
+    for gb in gb_list:
+        setup = model(params=params)
+        # Critical to this step is that setup.prepare_simulation() (specifically setup.create_grid())
+        # doesn't overwrite that we manually set a grid bucket to the model.
+        setup.gb = gb
+        pp.contact_conditions.set_projections(setup.gb)
```

Comment:
this only work for contact mechanics?

Reply:
I think we can generalize it to AbstractModel. Then you will need to implement a new model, say FlowModel, to use it. In practice, this will be a lightweight wrapper around your simulation script, which standardizes some names etc.

Reply:
To be clear, for this PR, I think we should generalize to AbstractModel (should be straightforward), but leave it at that.